File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/submodel/metamodel/connected/submodelelement/ConnectedSubmodelElementCollection.java
package org.eclipse.basyx.submodel.metamodel.connected.submodelelement;

import java.util.Map;

import org.eclipse.basyx.submodel.metamodel.api.reference.enums.KeyElements;
import org.eclipse.basyx.submodel.metamodel.api.submodelelement.ISubmodelElement;
import org.eclipse.basyx.submodel.metamodel.api.submodelelement.ISubmodelElementCollection;
import org.eclipse.basyx.submodel.metamodel.api.submodelelement.dataelement.IProperty;
import org.eclipse.basyx.submodel.metamodel.api.submodelelement.operation.IOperation;
import org.eclipse.basyx.submodel.metamodel.facade.SubmodelElementMapCollectionConverter;
import org.eclipse.basyx.submodel.metamodel.map.submodelelement.SubmodelElement;
import org.eclipse.basyx.submodel.metamodel.map.submodelelement.SubmodelElementCollection;
import org.eclipse.basyx.submodel.metamodel.map.submodelelement.dataelement.property.Property;
import org.eclipse.basyx.vab.modelprovider.VABElementProxy;

/**
 * "Connected" implementation of SubmodelElementCollection
 * 
 * @author rajashek
 *
 */
public class ConnectedSubmodelElementCollection extends ConnectedSubmodelElement implements ISubmodelElementCollection {
	public ConnectedSubmodelElementCollection(VABElementProxy proxy) {
		super(proxy);
	}

	@Override
	public Map<String, ISubmodelElement> getValue() {
		return getSubmodelElements();
	}

	@Override
	public boolean isOrdered() {
		return (boolean) getElem().getPath(SubmodelElementCollection.ORDERED);
	}

	@Override
	public boolean isAllowDuplicates() {
		return (boolean) getElem().getPath(SubmodelElementCollection.ALLOWDUPLICATES);
	}

	@Override
	public Map<String, ISubmodelElement> getSubmodelElements() {
		return ConnectedSubmodelElementFactory.getConnectedSubmodelElements(getProxy(), Property.VALUE, "");
	}

	@Override
	public Map<String, IProperty> getProperties() {
		return ConnectedSubmodelElementFactory.getProperties(getProxy(), Property.VALUE, "");
	}

	@Override
	public Map<String, IOperation> getOperations() {
		return ConnectedSubmodelElementFactory.getOperations(getProxy(), Property.VALUE, "");
	}
	
	@Override
	protected KeyElements getKeyElement() {
		return KeyElements.SUBMODELELEMENTCOLLECTION;
	}
	
	/**
	 * Get submodel element by given id
	 * @param id
	 * @return specific submodel element
	 */
	@SuppressWarnings("unchecked")
	@Override
	public ISubmodelElement getSubmodelElement(String id) {
		Map<String, Object> node =(Map<String, Object>) getProxy().getModelPropertyValue(id);
		ISubmodelElement element = ConnectedSubmodelElementFactory.getConnectedSubmodelElement(getProxy(), "", id, node);
		return element;			
	}

	/**
	 * Delete a submodel element by given id
	 * @param id
	 */
	@Override
	public void deleteSubmodelElement(String id) {
		getProxy().deleteValue(id);
	}
	
	/**
	 * adds a submodel element to the collection
	 * @param element
	 */
	@SuppressWarnings("unchecked")
	@Override
	public void addSubModelElement(ISubmodelElement element) {
		if (element instanceof SubmodelElement) {
			((SubmodelElement) element).setParent(getReference());

			// Convert "value" in SubmodelElementCollection from Map to Collection
			if (element instanceof SubmodelElementCollection) {
				Map<String, Object> converted = SubmodelElementMapCollectionConverter.smElementToMap((Map<String, Object>) element);
				getProxy().setModelPropertyValue(element.getIdShort(), converted);
				return;
			}
		}
		
		getProxy().setModelPropertyValue(element.getIdShort(), element);
	}
}

File: sdks/c++/basys.sdk.cc/src/submodel/submodel/connected/submodelelement/ConnectedSubmodelElementCollection.cpp
/*
 * ConnectedSubmodelElementCollection.cpp
 *
 *      Author: wendel
 */

#include <BaSyx/submodel/connected/submodelelement/ConnectedSubmodelElementCollection.h>
#include <BaSyx/submodel/api/submodelelement/property/IProperty.h>
#include <BaSyx/submodel/api/ISubModel.h>

namespace basyx {
namespace submodel {


ConnectedSubmodelElementCollection::ConnectedSubmodelElementCollection(std::shared_ptr<vab::core::proxy::IVABElementProxy> proxy):
  ConnectedSubmodelElement(proxy)
{}

void ConnectedSubmodelElementCollection::setValue(const basyx::specificCollection_t<ISubmodelElement> & value)
{
  //todo
  //this->setProxyValue(property::PropertyPaths::VALUE, value);
}

basyx::specificCollection_t<ISubmodelElement> ConnectedSubmodelElementCollection::getValue() const
{
  //todo 
  return basyx::specificCollection_t<ISubmodelElement>(); // this->getProxyCollection(property::PropertyPaths::VALUE);
}

void ConnectedSubmodelElementCollection::setOrdered(const bool & value)
{
  this->setProxyValue(SubmodelElementCollectionPaths::ORDERED, value);
}

bool ConnectedSubmodelElementCollection::isOrdered() const
{
  auto element = this->getProxy()->readElementValue(SubmodelElementCollectionPaths::ORDERED);
  return element.Get<bool>();
}

void ConnectedSubmodelElementCollection::setAllowDuplicates(const bool & value)
{
  this->setProxyValue(SubmodelElementCollectionPaths::ALLOWDUPLICATES, value);
}

bool ConnectedSubmodelElementCollection::isAllowDuplicates() const
{
  auto element = this->getProxy()->readElementValue(SubmodelElementCollectionPaths::ALLOWDUPLICATES);
  return element.Get<bool>();
}

void ConnectedSubmodelElementCollection::setElements(const basyx::specificMap_t<ISubmodelElement> & elements)
{
  //todo 
  //this->setProxyValue(SubmodelPaths::SUBMODELELEMENT, elements);
}

basyx::specificMap_t<ISubmodelElement> ConnectedSubmodelElementCollection::getElements() const
{
  //todo 
  
  //auto elements = this->getProxy()->readElementValue(SubmodelPaths::SUBMODELELEMENT);
  return basyx::specificMap_t<ISubmodelElement>();
}

}
}
